extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -288,6 +253,40 @@ export class ControllerNode extends ControllerTreeNode {
 	public get description(): string {
 		return super.description;
 	}
+
+	/**
+	 * Gets the TreeNode for the SQL master Instance. This is async since we need to fetch endpoint data
+	 * for the master instance first from the controller.
+	 */
+	public async getSqlMasterNode(): Promise<SqlMasterNode> {
+		if (!this._sqlMasterNodePromise) {
+			return this.createSqlMasterNode();
+		}
+		return this._sqlMasterNodePromise;
+	}
+
+	/**
+	 * Fetches the endpoint data for the SQL master instance, creates the TreeNode with that data and adds it
+	 * to the Endpoints folder node once complete. This is async since we need to fetch endpoint data for the
+	 * master instance first from the controller.
+	 */
+	private async createSqlMasterNode(): Promise<SqlMasterNode> {
+		this._sqlMasterNodePromise = new Promise<SqlMasterNode>(async (resolve, reject) => {
+			const response = await getEndPoints(this._clusterName, this._url, this._username, this._password, true);
+			if (response && response.endPoints) {
+				const master = response.endPoints.find(e => e.name && e.name === 'sql-server-master');
+				resolve(new SqlMasterNode(master.endpoint, this._endPointFolderNode, undefined, this.treeChangeHandler, master.description));
+			}
+			reject(new Error(localize('noEndpointsError', "Did not receive valid response when fetching endpoints.")));
+		});
+
+		this._sqlMasterNodePromise.then((sqlMasterNode) => {
```

Comment:
There are multiple places you're waiting on this promise, but don't handle the error case. This will result in unhandled promise errors. Similarly, having other nodes children get set arbitrarily on success of this will be flaky - it should be that on expansion it gets the content. Overall, this pattern looks wrong. Let's discuss